### PR TITLE
Update HPKE to latest draft

### PR DESCRIPTION
**Pull request by R1kM**
This PR updates the spec and implementation of HPKE to match the latest draft.
The PR is quite massive, but almost all changes are about updating the HPKE spec and adding spec tests, updating code/hpke accordingly, and refreshing dist/ for a large number of HPKE implementations and ciphersuites.
Changes not related to HPKE are:
- Bumping a rlimit in a file in code/ed25519
- Adding a Makefile target for Lib_Memzero0.ml that seems needed for HPKE spec tests
- Updating the specialization tactic in Meta.Interface to add the z3refresh option when typechecking rewritten definitions. This was needed to work around a known bug of Z3 (Overflow ... old vector) that consistently occured when verifying the full spliced HPKE in one go.

Updating the spec to the latest draft was done by @blipp and previously reviewed by @prosecco. I'll let them comment on what these changes entail, and what the spec tests cover. My only changes to this spec are to split key_schedule to make verification easier, and to add a few inline_for_extraction annotations.
Following the HACLxN methodology, and similar to the previous implementation, code in Hacl.Impl.HPKE is a verified, generic implementation of HPKE, which is parametric in the different underlying cryptographic primitives (DH, AEAD, Hash). This implementation is then specialized for a variety of ciphersuites in the different Hacl.HPKE.{DH}_{AEAD}_{Hash} modules in code/hpke.
The generic implementation (and hence the specializations and extracted code) currently only covers the Base mode of HPKE. Extending this to PSK, Auth, and PSKAuth could be done in a further PR.

CI is currently broken, but I was able to get consistent clean builds with the latest F* and KaRaMeL, i.e., F* version ac90dfc and KaRaMel version 5f1b47b

### CODE REVIEW

**Comment by karthikbhargavan** [comment]
This looks good, thanks! We should try to implement the other modes soon since much of the code for them is already embedded within the key schedule in this PR.

**Review by karthikbhargavan — COMMENTED** [review]
Do we have tests in the repo somewhere?

**Comment by R1kM** [comment]
Regarding tests, yes, we have spec tests added by @blipp in specs/tests/hpke.

Regarding other modes, I agree, a lot of the necessary code might already be in place. I'd like to add them in a subsequent PR though; this one is already quite massive, and I'd like to make sure it passes the CI that @W95Psp and @pnmadelaine have been working on, to ensure that the base mode at least is stable verification-wise.

**Review by karthikbhargavan — APPROVED** [review]
Agreed, let's merge this

**Comment by msprotz** [comment]
Aymeric, the Lib_Memzero0.ml addition is concerning. AFAIK, Karthik had managed to remove it in this commit: 94cb2431

The commit contains some info. Can you run a quick `git grep --no-index Lib_Memzero0 obj/*.ml`to figure out where this comes from? This is a spurious dependency from spec to lib which should not exist (why would specs depend on Low* code?).

File: specs/tests/hpke/hpke_test_vectors.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
+import json
```
Comment (msprotz) [inline specs/tests/hpke/hpke_test_vectors.py:1]: same remark here, this is really suboptimal -- Victor had done something in the past that would not need python but would rather read in OCaml the list of vectors, then return it to F* to iterate over so that the test could be written in F*... I'm concerned about the status of these tests

**Review by msprotz — CHANGES_REQUESTED** [review]
I have a variety of concerns that I listed in my review... happy to chat about this in greater detail and/or help out once I'm back from Oakland. Thanks!

File: code/meta/Meta.Interface.fst [hunk 2]
```diff
@@ -307,7 +307,7 @@ let rec visit_function (t_i: term) (st: state) (f_name: name): Tac (state & list
         let se_t = pack_sigelt (Sg_Let false [lb]) in
         let se_t = set_sigelt_quals [ NoExtract; Inline_for_extraction ] se_t in
         let se_t = match original_opts with
-          | Some original_opts -> add_check_with original_opts se_t
```
[the remark this answers is not in the document]
Comment (R1kM) [reply code/meta/Meta.Interface.fst]: Good point, I removed this and added additional z3refresh annotations in Hacl.Impl.HPKE

File: specs/tests/hpke/hpke_test_vectors.py (hunk 1 above)
Comment (R1kM) [reply specs/tests/hpke/hpke_test_vectors.py]: Benjamin indicated this file was not needed anymore, I removed it

**Comment by karthikbhargavan** [comment]
Do you think it would be possible to generate all the Hacl_HPKE_* files with a script? They look pretty boiler-plate right? We can leave this as a TODO for the future.

**Comment by blipp** [comment]
Replying to two comments by Jonathan concerning tests:

> can you comment on the main PR discussion as to why the tests were moved to hand-written ocaml?

Couple of reasons: The goal was to test the entire test vectors provided in a JSON by the RFC; I did not know how to do this in F* and was not aware of the previous efforts by Victor (see below); I was not aware that writing tests as much as possible in F* was the aspired goal for HACL* (i.e., I got the advice that writing tests in OCaml is fine).

> Victor had done something in the past that would not need python but would rather read in OCaml the list of vectors, then return it to F* to iterate over so that the test could be written in F*

Victor provided me with a link to his past efforts from approx. 3 years ago, they have not been merged: https://github.com/project-everest/hacl-star/compare/master...vdum_dev_merge_tests The concept looks great, and I believe it would be great if HACL* could adopt this style. It is hard for me to assess how long it would take to make this work with current F* and HACL*; regardless of this it would be some engineering effort to rewrite the tests completely in F*. Should I/we do this and wait with the merge until it is done?

**Comment by msprotz** [comment]
I don't think we should block the merge on moving tests to F* -- but we should block the merge on making sure tests are part of CI.

It's not a big issue to have tests in OCaml, but you have to keep in mind that:
- this leads to a proliferation of test styles each with their own helpers and styles
- the test is "further" away from the spec, so it might be more confusing to fix should the spec change
- this is another technology, language and set of libraries to master if someone wants to extend your work (F* + OCaml + python).

**Comment by R1kM** [comment]
I can look into adding tests to CI this weekend. @blipp , do you remember why the upper bounds on the AEAD spec were needed? It seems to be the last blocking part to merge this PR.

**Comment by blipp** [comment]
> I can look into adding tests to CI this weekend.

Thank you :)

>  @blipp , do you remember why the upper bounds on the AEAD spec were needed? It seems to be the last blocking part to merge this PR.

Working on it, should be able to give an answer today.

File: specs/Spec.Agile.AEAD.fsti [hunk 3]
```diff
@@ -119,7 +121,7 @@ let decrypted #a (c: cipher a) = p:plain a { S.length c = cipher_length p }
 // specs. Sigh.
 val encrypt: #(a: supported_alg) -> kv a -> iv a -> ad a -> p:plain a -> encrypted p
 val decrypt: #(a: supported_alg) -> kv a -> iv a -> ad a ->
-  c:cipher a { S.length c <= max_length a + tag_length a} ->
+  c:cipher a { S.length c <= cipher_max_length a} -> // TODO If my change of cipher can stay, then this check here is redundant
```
[the remark this answers is not in the document]
Comment (blipp) [reply specs/Spec.Agile.AEAD.fsti]: I am sorry for having left TODOs.

I want to make a case for keeping this change, i.e., adding the restriction to the `cipher` type (with my most recent commit removing TODO comments and simplifying the signature further). My reasoning in the following.

The type `plain` is already a bytestring with maximum length:

```
let plain (a: supported_alg) = s:S.seq uint8 { S.length s <= max_length a }
```

Thus, the `encrypt` function will only produce `cipher` values with some maximum length. However, the `cipher` type itself is not restricted, prior to this PR:

```
let cipher (a: supported_alg) = s:S.seq uint8 { S.length s >= tag_length a }
```

Furthermore, the `decrypt` function only takes a `cipher` value up to a certain maximum length:

```
val decrypt: #(a: supported_alg) -> kv a -> iv a -> ad a ->
  c:cipher a { S.length c <= max_length a + tag_length a} ->
```

This makes the “usability” of the type `cipher` somewhat limited: it contains bytestrings longer than the maximum allowed by `decrypt`, but these values cannot be produced by `encrypt` and also cannot be processed by `decrypt`.

Furthermore, in HPKE, while we can simply use `plain` as parameter in the outwards facing API of encryption functions:

```
val sealBase:
    cs:ciphersuite_not_export_only
  -> skE:key_dh_secret_s cs
  -> pkR:DH.serialized_point (kem_dh_of_cs cs)
  -> info:info_s cs
  -> aad:AEAD.ad (aead_alg_of cs)
  -> pt:AEAD.plain (aead_alg_of cs) ->
  Tot (option (key_dh_public_s cs & AEAD.encrypted #(aead_alg_of cs) pt))
```

we need to have the additional restriction required by `decrypt` in each function signature that does decryption:

```
val openBase:
    cs:ciphersuite_not_export_only
  -> enc:key_dh_public_s cs
  -> skR:key_dh_secret_s cs
  -> info:info_s cs
  -> aad:AEAD.ad (aead_alg_of cs)
  -> ct:AEAD.cipher (aead_alg_of cs) {Seq.length ct <= AEAD.max_length (aead_alg_of cs) + AEAD.tag_length (aead_alg_of cs)} ->
  Tot (option (AEAD.decrypted #(aead_alg_of cs) ct))
```

This can be done, but it is inconvenient: (1) it is necessary in the HPKE spec and in every spec that uses HPKE (2) it exposes the details of the AEAD inner structure (actual ciphertext plus tag) to the exterior.

By the way, in the comments of Spec.Agile.AEAD.fsti, the URL https://github.com/mitls/mitls-papers/wiki/Discussion-to-Converge-on-Design-Decisions is mentioned when talking about length restrictions of `plain`. This URL does not seem to be accessible anymore. GitHub throws an “Upgrade to GitHub Team or make this repository public to enable this feature.” error.

**Comment by blipp** [comment]
Please see this comment https://github.com/project-everest/hacl-star/pull/553#discussion_r888955792 concerning the changes in the AEAD spec.

The CI error we now get is because of the Mono->DotNet changes, I suppose. It is saying “mono not found”.

**Comment by msprotz** [comment]
Blipp I agree with your comment on AEAD.

**Comment by R1kM** [comment]
Thanks @blipp and @msprotz for the discussion. It seems we are reaching agreement on this PR, with the last remaining point being to include HPKE spec tests into the CI.
As far as I understand, there is consensus that HPKE spec tests should be moved to use F*/OCaml instead of Python but that this can be done in a subsequent PR once this one is merged. I'll open an issue as soon as this is done.
Am I missing something else?

FWIW, during today's F* meeting, we quickly discussed whether adding quasi-quotes to F* (https://github.com/FStarLang/FStar/pull/2598) could help with verification times for HACL* spec tests written in F*, cc @W95Psp who initiated this PR and was interested in this topic.

With #564 merged into master, I just merged master into this branch which should also restore CI and provide some guarantees before the final merge.

**Comment by msprotz** [comment]
>  Am I missing something else?

No it's good for me as long as those tests are on CI. I just don't want files in the source tree that are unused.

**Comment by R1kM** [comment]
I added the HPKE spec tests to the main Makefile, they are now executed when running `make test`.
I also slightly tweaked the main test file to return an error code when a test fails, similar to what other AutoTest modules are doing.
To do so, I had to disable tests for unsupported ciphersuites, creating the new `test-vectors-supported.json` file. Unsupported ciphersuites include the ones for which we do not currently have an executable spec (AES-GCM, P384, ...).
I also had to disable test cases using P256, key derivation with P256 leads for me to stack overflow issues, possibly due to a recursion going too deep? @Frosne Does this ring a bell/do you have spec tests for P256 somewhere?

**Comment by R1kM** [comment]
The output is the following (with the last lines repeated many times, omitted for clarity):
```
mode AuthPSK, Fatal error: exception Stack overflow
Raised by primitive operation at Z.sub in file "z.ml", line 62, characters 4-13
Called from Spec_P256_Lemmas.pow in file "obj/Spec_P256_Lemmas.ml", line 7, characters 22-41
Called from Spec_P256_Lemmas.pow in file "obj/Spec_P256_Lemmas.ml", line 7, characters 15-42
Called from Spec_P256_Lemmas.pow in file "obj/Spec_P256_Lemmas.ml", line 7, characters 15-42
```

**Comment by karthikbhargavan** [comment]
Spec.P256.Lemmas defines its own exponentiation [here](https://github.com/project-everest/hacl-star/blob/9332ebe6ede42a6db9d52363f8ccd95820e7081c/specs/ecdsap256/Spec.P256.Lemmas.fst#L20)

Presumably this is because the spec needs a more general exponentiation function than `pow2` 
However this function is very inefficient and should be replaced with binary exponentiation, otherwise we will keep hitting stackoverflows on spec tests.

**Comment by R1kM** [comment]
I don't know much about the P256 spec, but looking quickly, Spec.P256.Lemmas defines its exponentiation which is then used for normalization (inside `_norm` in Spec.P256) by calling modp_inv2_pow from `Spec.P256.Lemmas`.
The issue looks unrelated to this PR, do you think we should block on this for merging, or can we just open an issue to solve this at a later time?

**Comment by msprotz** [comment]
Sounds like we should merge HPKE and open a followup issue for P256. Some things to note on the followup issue:
- why is P256 essentially redefining its own copy of this: https://github.com/FStarLang/FStar/blob/master/ulib/prims.fst#L701
- is there a proof somewhere that prims.pow2 is equivalent to a binary one? surely, someone must have done this
- use that to replace calls to P256.pow2 with binary_pow2 + lemma call to maintain the proofs
- cross fingers that this works

**Review by msprotz — APPROVED** [review]
